src/libexpr/primops.cc

```diff
@@ -3927,20 +3922,26 @@ static void prim_replaceStrings(EvalState & state, const PosIdx pos, Value * * a
         bool found = false;
         auto i = from.begin();
         auto j = to.begin();
-        for (; i != from.end(); ++i, ++j)
+        size_t j_index = 0;
+        for (; i != from.end(); ++i, ++j, ++j_index)
             if (s.compare(p, i->size(), *i) == 0) {
                 found = true;
-                res += j->first;
+                auto v = cache.find(j_index);
+                if (v == cache.end()) {
+                    NixStringContext ctx;
+                    auto ts = state.forceString(**j, ctx, pos, "while evaluating one of the replacement strings passed to builtins.replaceStrings");
```

Comment:
I think in that case it would be better to use try/catch, so that we don't have to compute the dynamic string ahead of time and then let it go unused in almost all cases.

Reply:
I am conflicted whether to go with the try/catch or use something akin to `Result<T, E>` in Rust to propagate errors.

In any case, I think another PR should be made for this explicitly as it requires changing how the `force*` functions are called

Reply:
> `Result<T, E>`

I like that type, but that's not what any of the other code here uses, and I think it's for good reason. This is C++, so its native exceptions are a natural choice. It isn't really good at sum types.

> Edit: On further thought, I think if `errorCtx` parameter can be lazily evaluated, then nothing at the call site would need to change. However I don't know how to achieve this.

The same is achieved with an exception, without the overhead of creating a thunk.

> In any case, I think another PR should be made for this explicitly as it requires changing how the `force*` functions are called and other call sites can also benefit from adding position information.

I'm ok with incremental improvements.

Reply:
Sorry for the late reply. I have been experimenting with using variadic templates to send arguments into `addTrace` but this would involving moving a significant amount of functions in `eval.cc` to a header file like `eval-inline.hh`. I don't know if this is acceptable. In any case, I made a PR of my current progress for this change https://github.com/NixOS/nix/pull/8427.